[SRVKE-1271] Add kafka broker config map options (#50544)
openshift/openshift-docs: modules/serverless-kafka-broker-configmap.adoc

// Module included in the following assemblies:
//
// * /serverless/admin_guide/serverless-kafka-admin.adoc

:_content-type: PROCEDURE
[id="serverless-kafka-broker-configmap_{context}"]
= Configuring Kafka broker settings

> You can configure the number of topic partitions, replication factor, and bootstrap servers for a Kafka broker, by creating a config map and referencing this config map in the Kafka `Broker` object.

Author: It's not the number of bootstrap servers, it's the servers themselves :)
I did think about this, but I wanted to keep them in the order that they appear in the config example - do you think it's worth keeping them consistent with the order in the example, or should I still restructure this?

> You can configure the number of topic partitions, replication factor, and bootstrap servers for a Kafka broker, by creating a config map and referencing this config map in the Kafka `Broker` object.

Reviewer: It's far clearer to me if you place "the number of topic partitions" at the end of the list, in that case, but your call!

:FeatureName: Kafka broker
include::snippets/technology-preview.adoc[]

.Prerequisites

* You have cluster or dedicated administrator permissions on {product-title}.

> * The {ServerlessOperatorName}, Knative Eventing, and the `KnativeKafka` custom resource (CR) are installed on your {product-title} cluster.

Author: Yeah, tbh it reads weirdly without it IMO :) and we use "the" before it across all the other docs, so I'd rather keep it consistent.

> * You have created a project or have access to a project with the appropriate roles and permissions to create applications and other workloads in {product-title}.

Reviewer: ```suggestion
* You created a project or have access to a project that has the appropriate roles and permissions to create applications and other workloads in {product-title}.
```

Reviewer: ISG is very much for using "that" with restrictive clauses, and this pattern is found throughout our docs.

Peer review is ultimately advisory, though, so it's your call.

> * You have installed the OpenShift (`oc`) CLI.

Reviewer: ```suggestion
* You installed the OpenShift (`oc`) CLI.
```
Simpler tense.

Author: Wouldn't this then be inconsistent with other parts of the docs, e.g. above where it reads "You have created..."? I feel like "you installed" just reads strangely and wouldn't be consistent with the prereqs in all our other modules. Random example from our docs: https://docs.openshift.com/container-platform/4.11/serverless/develop/serverless-listing-event-sources.html#serverless-list-source-types-kn_serverless-listing-event-sources

I think this and the previous suggestions re wording would need to be raised as a suggestion for our docs as a whole, rather than changed for this one module, since they apply to a whole bunch of places. If there's anywhere in the guidelines where these are suggested and you could share links to those it would be helpful for future planning though.

Reviewer: There isn't consistency in the docs themselves:
- https://docs.openshift.com/container-platform/4.11/machine_management/user_infra/adding-aws-compute-user-infra.html#prerequisites_adding-aws-compute-user-infra
- https://docs.openshift.com/container-platform/4.11/updating/updating-clusters-sro.html#updating-running-a-preflight-check-for-sro_updating-clusters-sro
- https://docs.openshift.com/container-platform/4.11/registry/configuring_registry_storage/configuring-registry-storage-baremetal.html#registry-configuring-registry-storage-rhodf-cephrgw_configuring-registry-storage-baremetal

ISG is explicit about using simple grammar whenever possible. :shrug:

Reviewer: SSG, meanwhile, offers the simplest possible present tense alternative:

> JDK 11 or later is installed.

https://redhat-documentation.github.io/supplementary-style-guide/#prerequisites

Author: > There isn't consistency in the docs themselves
I specifically meant the serverless docs, since that's the product I work on :)
Thanks for the info!

.Procedure

. Modify the `kafka-broker-config` config map, or create your own config map that contains the following configuration:
+
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: <config_map_name> <1>
  namespace: <namespace> <2>
data:
  default.topic.partitions: <integer> <3>
  default.topic.replication.factor: <integer> <4>
  bootstrap.servers: <list_of_servers> <5>
----
<1> The config map name.
<2> The namespace where the config map exists.
<3> The number of topic partitions for the Kafka broker. This controls how quickly events can be sent to the broker. A higher number of partitions requires greater compute resources.
<4> The replication factor of topic messages. This prevents against data loss. A higher replication factor requires greater compute resources and more storage.
<5> A comma separated list of bootstrap servers. This can be inside or outside of the {product-title} cluster, and is a list of Kafka clusters that the broker receives events from and sends events to.
+
[IMPORTANT]
====

> The `default.topic.replication.factor` value must be less than or equal to the number of Kafka broker instances in your cluster. For example, if you only have one Kafka broker, the `default.topic.replication.factor` value should not be more than 1.

Reviewer: :warning: ISG
```suggestion
The `default.topic.replication.factor` value must be less than or equal to the number of Kafka broker instances in your cluster. For example, if you only have one Kafka broker, the `default.topic.replication.factor` value should not be more than `"1"`.
```
Or "one."

:FeatureName: Kafka broker
include::snippets/technology-preview.adoc[]

.Prerequisites

* You have cluster or dedicated administrator permissions on {product-title}.
* The {ServerlessOperatorName}, Knative Eventing, and the `KnativeKafka` custom resource (CR) are installed on your {product-title} cluster.
* You have created a project or have access to a project with the appropriate roles and permissions to create applications and other workloads in {product-title}.
* You have installed the OpenShift (`oc`) CLI.

.Procedure

. Modify the `kafka-broker-config` config map, or create your own config map that contains the following configuration:
+
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: <config_map_name> <1>
  namespace: <namespace> <2>
data:
  default.topic.partitions: <integer> <3>
  default.topic.replication.factor: <integer> <4>
  bootstrap.servers: <list_of_servers> <5>
----
<1> The config map name.
<2> The namespace where the config map exists.
<3> The number of topic partitions for the Kafka broker. This controls how quickly events can be sent to the broker. A higher number of partitions requires greater compute resources.

> <4> The replication factor of topic messages. This prevents against data loss. A higher replication factor requires greater compute resources and more storage.

Reviewer 2: I'd add somewhere the note that is also upstream https://knative.dev/docs/eventing/brokers/broker-types/kafka-broker/#configure-a-kafka-broker:

> The default.topic.replication.factor value must be less than or equal to the number of Kafka broker instances in your cluster. For example, if you only have one Kafka broker, the default.topic.replication.factor value should not be more than 1.

Author: Added, thank you!

  bootstrap.servers: <list_of_servers> <5>
----
<1> The config map name.
<2> The namespace where the config map exists.
<3> The number of topic partitions for the Kafka broker. This controls how quickly events can be sent to the broker. A higher number of partitions requires greater compute resources.
<4> The replication factor of topic messages. This prevents against data loss. A higher replication factor requires greater compute resources and more storage.
<5> A comma separated list of bootstrap servers. This can be inside or outside of the {product-title} cluster, and is a list of Kafka clusters that the broker receives events from and sends events to.
+
[IMPORTANT]
====
The `default.topic.replication.factor` value must be less than or equal to the number of Kafka broker instances in your cluster. For example, if you only have one Kafka broker, the `default.topic.replication.factor` value should not be more than 1.
====
+
.Example Kafka broker config map
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: kafka-broker-config
  namespace: knative-eventing
data:
  default.topic.partitions: "10"
  default.topic.replication.factor: "3"

>   bootstrap.servers: "my-cluster-kafka-bootstrap.kafka:9092"

Reviewer 3: This config isn't enough to deploy Broker. Created broker fails with:

> _ChannelTemplateFailed: Error on setting up the ChannelTemplate: empty or missing value for config_

Reviewer 2: I think Chris means that we need to explicitly specify the broker class down at line 76
```
    # case-sensitive
    eventing.knative.dev/broker.class: Kafka
```

Reviewer 2: That's how I read it as well but I think extra clarification is very well appreciated by users reading our docs, thank you!

<5> A comma separated list of bootstrap servers. This can be inside or outside of the {product-title} cluster, and is a list of Kafka clusters that the broker receives events from and sends events to.
+
[IMPORTANT]
====
The `default.topic.replication.factor` value must be less than or equal to the number of Kafka broker instances in your cluster. For example, if you only have one Kafka broker, the `default.topic.replication.factor` value should not be more than 1.
====
+
.Example Kafka broker config map
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: kafka-broker-config
  namespace: knative-eventing
data:
  default.topic.partitions: "10"
  default.topic.replication.factor: "3"
  bootstrap.servers: "my-cluster-kafka-bootstrap.kafka:9092"
----

. Apply the config map:
+
[source,yaml]
----
$ oc apply -f <config_map_filename>
----

. Specify the config map for the Kafka `Broker` object:
+
.Example Broker object
[source,yaml]
----
apiVersion: eventing.knative.dev/v1
kind: Broker
metadata:
  name: <broker_name> <1>
  namespace: <namespace> <2>

>   eventing.knative.dev/broker.class: Kafka <3>

Reviewer 3: ```suggestion
  annotations:
    eventing.knative.dev/broker.class: Kafka <3>
```

This should be an annotation, same as here: https://knative.dev/docs/eventing/configuration/broker-configuration/#channel-implementation-options
